src/interrupt/mod.rs

```diff
@@ -0,0 +1,203 @@
+//! pin-based interrupts and MSI interrupts.
+//!
+//! A device may support multiple types of interrupts, and each type of interrupt may support one
+//! or multiple interrupt sources. For example, a PCI device may support:
+//! * Legacy Irq: exactly one interrupt source.
+//! * PCI MSI Irq: 1,2,4,8,16,32 interrupt sources.
+//! * PCI MSIx Irq: 2^n(n=0-11) interrupt sources.
+//!
+//! A distinct Interrupt Source Identifier (ISID) will be assigned to each interrupt source.
+//! An ID allocator will be used to allocate and free Interrupt Source Identifiers for devices.
+//! To decouple the vm-device crate from the ID allocator, the vm-device crate doesn't take the
+//! responsibility to allocate/free Interrupt Source IDs but only makes use of assigned IDs.
+//!
+//! The overall flow to deal with interrupts is:
+//! * The VMM creates an interrupt manager
+//! * The VMM creates a device manager, passing on an reference to the interrupt manager
+//! * The device manager passes on an reference to the interrupt manager to all registered devices
+//! * The guest kernel loads drivers for virtual devices
+//! * The guest device driver determines the type and number of interrupts needed, and update the
+//!   device configuration
+//! * The virtual device backend requests the interrupt manager to create an interrupt group
+//!   according to guest configuration information
+
+use std::fs::File;
+use std::sync::Arc;
+
+/// Reuse std::io::Result to simplify interoperability among crates.
+pub type Result<T> = std::io::Result<T>;
+
+/// Data type to store an interrupt source identifier.
+pub type InterruptIndex = u32;
+
+/// Data type to store an interrupt source type.
+#[derive(Copy, Clone, Debug)]
+pub enum InterruptSourceType {
+    /// Legacy Pin-based Interrupt.
+    /// On x86 platforms, legacy interrupts are routed through 8259 PICs and/or IOAPICs.
+    LegacyIrq,
+    /// Message Signaled Interrupt (PCI MSI/PCI MSIx).
+    /// Some non-PCI devices (like HPET on x86) make use of generic MSI in platform specific ways.
+    PciMsiIrq,
+}
+
+/// Trait to manage interrupt sources for virtual device backends.
+///
+/// The InterruptManager implementations should protect itself from concurrent accesses internally,
+/// so it could be invoked from multi-threaded context.
+pub trait InterruptManager {
+    /// Create an [InterruptSourceGroup](trait.InterruptSourceGroup.html) object to manage
+    /// interrupt sources for a virtual device
+    ///
+    /// An [InterruptSourceGroup](trait.InterruptSourceGroup.html) object manages all interrupt
+    /// sources of the same type for a virtual device.
+    ///
+    /// # Arguments
+    /// * interrupt_type: type of interrupt source.
+    /// * base: base Interrupt Source ID to be managed by the group object.
+    /// * count: number of Interrupt Sources to be managed by the group object.
+    fn create_group(
+        &mut self,
+        interrupt_type: InterruptSourceType,
```

Comment:
I think this should not be in the InterruptManager trait. In fact I don't think InterruptSourceType should be in this trait at all. Rather, there should be multiple implementations of the InterruptManager, for example one that returns an InterruptSourceGroup for legacy interrupts and one that returns an InterruptSourceGroup for MSIs. I would make this something like

    trait InterruptManager {
        type Config;
        // use interior mutability, more than one device will own the InterruptManager
        fn create_source(&self, c: Config) -> Result<Box<dyn InterruptSourceGroup>>;
    }

A PCI host bridge will have two interrupt managers, one for legacy interrupts and one for MSIs. In turn the one for legacy interrupts could delegate to the GSI (8259/IOAPIC) interrupt manager.

Reply:
This proposal has two drawbacks:
1) carrying on multiple interrupt managers
2) multiple interrupt managers need cooperation on the global KVM routing table.
We still need another level of wrapper manager to dispatch different type of requests to different InterruptManager.

With the "interrupt_type: InterruptSourceType" parameter, we may build a interrupt manager to handle both legacy and msi interrupt. Or we could be a hierarchy as
struct InterruptManagerWrapper {
        struct InterruptManagerLegacy,
        struct InterruptManagerMsi,
}.
By this way, the client side is simpler.